UI: Alloc FS: File Viewer

DingoEatingFuzz opened the pull request:
Completes the feature set of the allocation explorer by making files viewable. File viewing currently supports:

1. Text files (logs, json, code, etc.)
2. Images (jpg, png, svg, webp, gif, etc.)

If a file is not supported by the viewer, there will still be a link to open the API endpoint in a new tab for convenience. Depending on the file, the browser will either embed it (e.g., pdf, video) or download it for you (arbitrary binary).

![image](https://user-images.githubusercontent.com/174740/62197671-22344400-b335-11e9-8385-a0ae0d3822b1.png)
_A text file shown in the file cat view_

![image](https://user-images.githubusercontent.com/174740/62197742-46902080-b335-11e9-9c42-b019ea7f4b44.png)
_Tailing a file in the file viewer_

![image](https://user-images.githubusercontent.com/174740/62197843-6f181a80-b335-11e9-98f0-e141b1ce0e41.png)
_An embedded image from within an allocation_

![image](https://user-images.githubusercontent.com/174740/62198185-0b422180-b336-11e9-8690-208fe062b3ad.png)
_An unsupported file type_

johncowen commented on ui/tests/integration/image-file-test.js:
```diff
@@ -0,0 +1,111 @@
+    alt: 'This is the alt text',
+    size: 123456,
+  };
+
```
Would using a data-uri or a Blob here? Loading an svg using a data-uri would be the simplest I think. I'm not sure what you need to do exactly here so I might be going down the wrong path - but I think a datauri would work for what you need..maybe..?

DingoEatingFuzz replied:
Ah cool, thank you both for investigating. I assumed poorly that a data-uri wouldn't trigger the onload event, but in hindsight it makes sense that the API is consistently abstract from the data scheme.

backspace commented on ui/tests/unit/adapters/node-test.js:
```diff
@@ -20,7 +20,6 @@ module('Unit | Adapter | Node', function(hooks) {
     this.server.create('allocation', { id: 'node-1-2', nodeId: 'node-1' });
     this.server.create('allocation', { id: 'node-2-1', nodeId: 'node-2' });
     this.server.create('allocation', { id: 'node-2-2', nodeId: 'node-2' });
-    this.server.logging = true;
```
I was wondering about this!

DingoEatingFuzz replied:
This deletion was special enough to get its own commit 🤓 

It was already in the code (as evident by the diff) and I just happened to notice it spewing log lines in the test runner. 

If you aren't familiar with `this.server.logging = true`, it logs every HTTP request that goes through Pretender. Great for debugging, super noisy though.

backspace replied:
yeah, what do you think of [this kind of solution](https://github.com/backspace/prison-rideshare-ui/blob/primary/mirage/config.js#L7) that lets you add a query parameter to turn on logging? I like it more than having to edit code 😆

DingoEatingFuzz replied:
ooOOOoo, that's quite neat indeed. I like that. And nothing about the qunit runner removes that query param in an annoying manner?

backspace replied:
I haven’t had any trouble with it so far! 🎉

johncowen replied:
+1 to doing this. We do this in consul lots, but we use `document.cookies` instead. Using cookies means settings stick around when the URL changes, which is handy. Then we use Web Inspector to edit the cookies, which turns the Web Inspector into kind of a semi 'dev preferences' panel.

Actually, we didn't ever add a setting to turn HTTP logging on and off (we use it more to control mocks/doubles), but I think I'd add that also as we have some noisy logging going on aswell which I don't mind but other folks might.

backspace commented:
oh yes, one other thing I meant to include, is that I think the spacing in the breadcrumbs looks a bit off when viewing a file vs a directory

<img width="526" alt="ls-spacing" src="https://user-images.githubusercontent.com/43280/62378014-7abc3a80-b509-11e9-8ff0-7a670c666800.png">

<img width="526" alt="file-spacing" src="https://user-images.githubusercontent.com/43280/62378031-827bdf00-b509-11e9-94d9-1208c7db76eb.png">

I’m not sure what’s preferable: make the container the same height for both to avoid jumping when browsing from a directory into a file, or have the text move down when it’s a file…??

johncowen commented:
Hey folks the other thing I've noticed here (thanks for the grabs @backspace !), there's a lot of testing for emojis going on here, which I believe are URL safe (yet multibyte).

Are there tests for paths/filenames with things like spaces and question marks, I can't see anything here? I also had a quick look in the other PRs related to this but I didn't see anything (but I might have missed it).

Is this something being tested for elsewhere?

John

DingoEatingFuzz commented on ui/app/templates/components/image-file.hbs:
```diff
@@ -0,0 +1,11 @@
+  <span class="image-file-caption-primary">
+    <strong data-test-file-name>{{fileName}}</strong>
+    {{#if (and width height)}}
+      <span data-test-file-stats>({{width}}px &times; {{height}}{{#if size}}px, {{format-bytes size}}{{/if}})</span>
```
Yep 🙈

DingoEatingFuzz commented on ui/app/styles/components/image-file.scss:
```diff
@@ -0,0 +1,26 @@
+    display: block;
+    color: $grey-light;
+    font-style: italic;
+  }
```
Indeed. Will delete.

DingoEatingFuzz commented on ui/app/templates/components/task-file.hbs:
```diff
@@ -0,0 +1,39 @@
+  {{else}}
+    <div data-test-unsupported-type class="empty-message is-hollow">
+      <h3 class="empty-message-headline">Unsupported File Type</h3>
+      <p class="empty-message-body message">The Nomad UI could not render this file, but you can still call view the file directly.</p>
```
Um, how it got there is lost to time, but I'll remove it 🔪

DingoEatingFuzz commented on ui/app/components/streaming-file.js:
```diff
@@ -0,0 +1,96 @@
+    // Force the scroll position to the bottom of the window when starting streaming
+    this.logger.one('tick', () => {
+      run.scheduleOnce('afterRender', () => this.synchronizeScrollPosition(true));
+    });
```
I agree, and I appreciate your attention to detail.

DingoEatingFuzz commented on ui/app/templates/allocations/allocation/task/fs.hbs:
```diff
@@ -1,31 +1,45 @@
+    {{else}}
+      <div class="fs-explorer boxed-section">
+        <div class="boxed-section-head">
+          <nav class="breadcrumb" data-test-fs-breadcrumbs>
```
Yeah, I thought the same thing, then thought that duplication isn't that bad, then decided I didn't want to touch anything I didn't need to...

...then did way more than I needed to in Mirage.

DingoEatingFuzz commented on ui/app/templates/components/task-file.hbs:
```diff
@@ -0,0 +1,39 @@
+    <h3 class="title is-4">Cannot fetch file</h3>
+    <p>The files for this task are inaccessible. Check the condition of the client the allocation is on.</p>
+  </div>
+{{/if}}
```
Yeah, this is definitely gross. I'll look into it.

DingoEatingFuzz commented on ui/app/templates/components/task-file.hbs:
```diff
@@ -0,0 +1,39 @@
+    {{#if isStreamable}}
+      <button data-test-log-action="toggle-stream" class="button is-white" onclick={{action "toggleStream"}}>
+        {{x-icon (if logger.isStreaming "media-pause" "media-play") class="is-text"}}
+      </button>
```
Interesting, must be some sort of `width:auto` situation combined with a long log line? I'll look into this too.

DingoEatingFuzz commented:
Ooookay, all comments addressed!

Everything is in new commits, so that's probably the easiest way to re-review, but I'll go over some of the bigger things.

- **Errant No Connection error**: This was due to the allocation's client not being loaded in time, which I fixed and added a test for.
- **File cat can go off screen**: This was some CSS that was oddly related to the clipped tooltip `overflow-x:auto` bug from a couple months ago. I set the width while accounting for the margin and it's good now.
- **Breadcrumbs have awkward bottom-margin when viewing a file**: This was due to the button height, which was no secret. I fixed it with an existing style that I forget existed: `is-compact`. This modifier also needed to be tweaked since this is the first time it's being used for multiple buttons in a bar.
- **Missing test coverage for tricky characters like "?"**: I added a new set of possible characters to the mirage factory. Everything is still working fine!

johncowen commented:
Hey @DingoEatingFuzz ,

I gave the URL chars a try, some work, but I could definitely get things to go awry with `%`'s at least. I could reasonably consistently get a 'URI malformed' to appear.

<img width="705" alt="Screenshot 2019-08-09 at 09 25 48" src="https://user-images.githubusercontent.com/554604/62782126-df533880-bab9-11e9-9a25-e1b74cad8d4e.png">


Don't let this hold you up though, if you want to move on with this and look at that later.

John

backspace commented on ui/app/components/task-file.js:
```diff
@@ -0,0 +1,146 @@
+          }
+          return response;
+        },
+        error => this.nextErrorState(error)
```
Nice fix! 🤩

backspace commented on ui/app/templates/allocations/allocation/task/fs.hbs:
```diff
@@ -1,31 +1,15 @@
+    {{else}}
+      <div class="fs-explorer boxed-section">
+        <div class="boxed-section-head">
+          {{fs-breadcrumbs task=task path=path}}
```
Extraction success! 😀

backspace commented on ui/tests/acceptance/task-fs-test.js:
```diff
@@ -256,12 +296,40 @@ module('Acceptance | task fs', function(hooks) {
+      secondAttempt.url.split('?')[0],
+      `/v1/client/fs/readat/${allocation.id}`,
+      'Server is hit second'
+    );
```
nice additions!

backspace commented on ui/tests/acceptance/task-fs-test.js:
```diff
@@ -256,12 +296,40 @@ module('Acceptance | task fs', function(hooks) {
   test('viewing a file', async function(assert) {
+    const node = server.db.nodes.find(allocation.nodeId);
+
+    server.logging = true;
```
Is this meant to be left here?

DingoEatingFuzz replied:
Hahaha, this is extra funny considering I'm removing it elsewhere in the same PR.

DingoEatingFuzz commented:
One last re-review? 🤞 

I addressed the character encoding issue and fixed the flapping tests (which were due to illegal characters I added to further test the encoding issue).

backspace commented on ui/out:
```diff
@@ -0,0 +1,1317 @@
+Environment: test
```
Looks like a stray inclusion of this file?

DingoEatingFuzz replied:
Lol, I can't win.
